src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    return services_list_changed(m);
+}
+
+
+
+static int
+_manager_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+    char *str;
+    int r;
+
+    pending->manager_slot = sd_bus_slot_unref(pending->manager_slot);
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    do {
+
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (strcmp(str, "State") == 0) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
+        r = sd_bus_message_exit_container(m);
+
+    return r;
+}
+
+static int
+dbus_connection_get_manager_properties(void)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sd_bus_call_method_async(bus, &_ctx.manager_slot,
+        "net.connman", "/", "net.connman.Manager", "GetProperties",
+        _manager_properties_changed, &_ctx, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+dbus_connection_get_service_properties(void)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sd_bus_call_method_async(bus, &_ctx.service_slot,
+        "net.connman", "/", "net.connman.Manager", "GetServices",
+        _services_properties_changed, &_ctx, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+SOL_API enum sol_connman_state
+sol_connman_get_state(void)
+{
+    return _ctx.connman_state;
+}
+
+static int
+_set_state_property_changed(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+
+    pending->state_slot = sd_bus_slot_unref(pending->state_slot);
+
+    return sol_bus_log_callback(reply, userdata, ret_error);
+}
+
+SOL_API int
+sol_connman_set_offline(bool enabled)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sd_bus_call_method_async(bus, &_ctx.state_slot, "net.connman", "/",
+        "net.connman.Manager", "SetProperty", _set_state_property_changed,
+        &_ctx, "sv", "OfflineMode", "b", enabled);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+SOL_API bool
+sol_connman_get_offline(void)
+{
+    if (_ctx.connman_state != SOL_CONNMAN_STATE_OFFLINE)
```

Comment:
no... this should be the `OfflineMode` property. I understand this is confusing, so should we rename the `get_offline()` and `set_offline()` to `set_radios_offline()`? Or `set_airplane_mode()`?